### ISSUE TITLE
Commitlog space amplification in sync mode

### ISSUE THREAD

**Issue by xemul** [scylladb/scylla#7663]
As was found out during LWT stress test research, when commitlog works in sync mode it writes many aligned entries on disk (e.g. ~200 bytes of a paxos record is padded up to 4K and is written on disk) and this causes the commitlog size to grow much faster, than it's expected by the rest of the system.

Since memtables grow slower and pin individual segments in memory, commitlog runs out of its allowed space and continues growing by allocating new and new segments slowing things down heavily. The time dirty manager or commitlog flush timer decides to flush memtables and release the segments at some point, but that's rather late already.

There are two things that can be done here:

1. Reduce the record size to 512 bytes. Likely filesystem will work OK even with that size.
2. Do not flush each small record immediately, but try to collect several of those. This will make the commitlog sync latency larger, but shouldn't spoil the LWT latency that much (or at all).

**Comment by elcallio** [comment]
Batch mode commitlog already "collects" whatever incoming writes appear while a disk write is being done. For it to have any meaningful impact though, query concurrency needs to be high. 
If you mean _always_ wait for X bytes to come in, you could end up in a situation where you wait forever. Timeout, sure, but that seems weird...
A better solution in that case would be to write the buffer, but save the slack - <disk block size> last as "overlap", and on next flush overwrite the last block on disk + the new blocks. The big problem with that is you might end up breaking an entry you previously promised was safe.

**Comment by avikivity** [comment]
> Batch mode commitlog already "collects" whatever incoming writes appear while a disk write is being done. For it to have any meaningful impact though, query concurrency needs to be high.
> If you mean _always_ wait for X bytes to come in, you could end up in a situation where you wait forever. Timeout, sure, but that seems weird...

Why is it weird? Optimize for throughput, while not breaking latency bounds.

> A better solution in that case would be to write the buffer, but save the slack - last as "overlap", and on next flush overwrite the last block on disk + the new blocks. The big problem with that is you might end up breaking an entry you previously promised was safe.

The other big problem is that the write amplification is not gone, only the space amplification. In addition the FTL does not overwrite, it always allocates new space, so it will have to do garbage collection to free the overwritten blocks.

**Comment by elcallio** [comment]
But having records wait for other records, then to flush "when X entries wait or Y seconds have passed" is more or less just doing periodic writes. The only difference would be the "I did actually write to disk before moving on". If we assume the LWT-esque write mode is always/mostly the case, maybe. But if it is (relatively) rare, why not just force-flush before moving on for these writes? Or add a waitable return type that signals when stuff is on disk, and then it is just periodic, but caller can opt to wait.

Even just reducing write/flush period (or the "timeout" above) will increase write amp. You will write more often if you write more often (hello tautology!).

**Comment by kostja** [comment]
@avi we should measure the impact of garbage collection. The scenario is not unusual and I don't see why not let the controller do its job. 
I'm more concerned with @elcallio comment that we should make sure the block we overwrite is not corrupt. @elcallio I think the risk of corruption is present even after a controller returns success, we can still get the garbage back on read. But we should be able to detect a corruption happened with block checksums. InnoDB uses a doublewrite buffer (*) for this purpose: the last block being written is always first written to a separate location, and then flushed to the database.
With commit log, we could have a separate area where we write out "incomplete" blocks, perhaps 16 block long. Once this area is filled up, or once we have enough data to fill a full block, we append it to the actual segment.
(*) https://dev.mysql.com/doc/refman/5.7/en/innodb-doublewrite-buffer.html

**Comment by elcallio** [comment]
If you cannot trust a written page is actually ok after controller ack, how can you trust two pages written?
That being said, yes, of course you can double-buffer overlapping pages. The tricky part is maybe
a.) knowing a page was actually borked - since we write and crc check individual entries, it is not inconceivable to mistake an unfilled page from a corrupted on, and maybe decide on selecting the "wrong" double-page for recover.
b.) how do you bookkeep this in a reasonable fashion? I mean tracking small-page cache, correlate to actual segment and segment position - and do so in a way that is hopefully as much or even more safe than the commitlog it is securing? Looks like a lot of data being written + flushed to keep track of that page.
Point being that the current log format/scheme might not translate well here - it is after all just a clone of the original origin mechanism.

**Comment by kostja** [comment]
> If you cannot trust a written page is actually ok after controller ack, how can you trust two pages written?

The point of writing two pages is not increased trust, is being able to fall back to the first write in case of a failure of the second. The second write is also used to verify the second: once written, you can read back the content and compare it.

> That being said, yes, of course you can double-buffer overlapping pages. The tricky part is maybe
> a.) knowing a page was actually borked - since we write and crc check individual entries, it is not inconceivable to mistake an unfilled page from a corrupted on, and maybe decide on selecting the "wrong" double-page for recover.

Yes, checksuming individual entries has to be replaced with checksuming writes. 

> b.) how do you bookkeep this in a reasonable fashion? I mean tracking small-page cache, correlate to actual segment and segment position - and do so in a way that is hopefully as much or even more safe than the commitlog it is securing? Looks like a lot of data being written + flushed to keep track of that page.

If we discuss the specific solution I suggested, with extra pages used to collect enough content for a compact write, we could store them at segment end and always read after the sequential segment content on recovery. The pages in the back should reflect the page number of the page in the segment they represent.

> Point being that the current log format/scheme might not translate well here - it is after all just a clone of the original origin mechanism.

Perhaps.

**Comment by elcallio** [comment]
How do you determine which block in the file the "extra, last" blocks map to? Replace file format encoding to be per-page, with header (with page index) + checksum for every disk block? This would effectively require a page-aware fragmented buffer that does all this (and hopefully uses data that is still relevant once written) when serializing to internal buffer...
Sounds more and more like this would be the triumphant return of using good old-fashioned memory mapping! :-)

**Comment by kostja** [comment]
@elcallio I think we can discuss the details in an RFC, but yes, we do need to checksum physical writes, and since the smallest write is a page, it means checksumming is per-page or per multiple pages. I'm not an expert in C++ and don't know what it would require if we talk using modern C++ concepts.

**Comment by avikivity** [comment]
> But having records wait for other records, then to flush "when X entries wait or Y seconds have passed" is more or less just doing periodic writes. The only difference would be the "I did actually write to disk before moving on".

And what a difference that is!

> If we assume the LWT-esque write mode is always/mostly the case, maybe.

No, we don't.

> But if it is (relatively) rare, why not just force-flush before moving on for these writes?

Because that causes huge space amplification. Or did you mean something else?

> Or add a waitable return type that signals when stuff is on disk, and then it is just periodic, but caller can opt to wait.
> 
> Even just reducing write/flush period (or the "timeout" above) will increase write amp. You will write more often if you write more often (hello tautology!).

There is strong similarity to periodic, but the two approaches are from different directions. Batch+wait is: it has to be safe, but let's sacrifice some minimum latency for throughput. Periodic is: safety is more or less optional, we give 10 seconds of bad data, but if throughput happens to force a write to disk it may be a little less.

If the implementations converge, great.

**Comment by avikivity** [comment]
@kostja yes, that's an extra risk with overwrites, you can end up in a situation where you have neither the before-write data or the after-write data.

It's less likely on SSD since the storage writes to a new place and updates a pointer, but you have to trust the controller.

IMO another reason not to do overwrites but to do appends.

**Comment by elcallio** [comment]
@avikivity why not add an optional timeout to adding CL entry that waits for maximum X time for a flush to happen organically, or else forces it with the smallest possible space waste? Basically a slightly more latency-generating batch mode. Only difference being that the flushing still happens in the periodic task, the timeout just makes up push up next period a bit, and allows us waiting for a given position to be flushed?

**Comment by avikivity** [comment]
Sounds reasonable. So flush happens when we reach acceptable space amplification, or the timeout, whichever happens first.

Note that if you accumulate enough data for 1.4 blocks, you'll want to flush just the first block to avoid the extra amplification that comes from writing two (unless the data in the extra .4 block is forced by timeout).

**Comment by kostja** [comment]
This patch rebased against mater: https://github.com/kostja/scylla/tree/commitlog-cleanup

**Comment by raphaelsc** [comment]
> This patch rebased against mater: https://github.com/kostja/scylla/tree/commitlog-cleanup

@kostja: As of https://github.com/torvalds/linux/commit/eda77982729b7, XFS serializes all DIO unaligned to fs block size. So unfortunately we can't align to sector size given that it's smaller than default fs blk size. I think the actual solution will be either patching Linux XFS to avoid serialization of sub-block write if underlying extent is already written, or tweaking our setup scripts to format XFS with blk size of 512.  We'll need to patch seastar here https://github.com/scylladb/seastar/blob/master/src/core/file.cc#L105 as it expects a minimum write alignment for 4096 bytes.

**Comment by kostja** [comment]
Correct, the patch uses a patched seastar, the seastar branch is mentioned in the commit comment.
I guess given the substandard traetment by XFS of 512 byte writes, we need have to abandon the idea of switching to smaller write block size in the commit log and explore other alternatives. I agree with Avi expecting 512 byte XFS block size in all scylla setups is unreasonable, at least today.

**Comment by avikivity** [comment]
We run the mkfs ourselves, so we can improve new deployments. See dist/common/scripts/scylla_raid_setup.

**Comment by avikivity** [comment]
(and the non-512 bytes setup will work, just not as well; we detect the block size)

**Comment by kostja** [comment]
Regarding this one, I believe we changed Scylla setup to use 512 byte XFS blocks where possible. @elcallio is there something on your todo list that can further improve the matter?

**Comment by elcallio** [comment]
Afaik, as long as we get correct info from seastar/file, we should use as small blocks as possible, i.e. minimize disk waste. Beyond complete rethink of disk file layout, I don't think there is more to do here.

**Comment by avikivity** [comment]
We use 1k blocks, see 6977064693894202024b510bcd2c2ee375c8e177.
